### ISSUE TITLE
Scoring based on recorded Actions

### ISSUE THREAD

**Issue by mgale456** [uiuc-mrdc/cogs#10]
I already started on this with a function in the GameParticipant model that counts the total for a ScoringType and returns it with the multiplier.
Still need to translate these into a score and save it in the Db.

Don't forget we need to account for nonlinear point values (brewing potions this year). One way to do this would be to have a custom function if ScoringType.value == 0, otherwise just multiply by ScoringType.value for linear scoring.

Also, it would be good to rearrange the counting function a bit so that it correctly only queries the Db once per call of the Score function. I think this should be possible, since Django caches the querysets.

**Comment by mgale456** [comment]
Scoring works, but has one edge case that I know of. For the counters, counting up, changing multiplier, and counting down results in the wrong total score.

**Comment by mgale456** [comment]
TLDR: desynced up and down db tuples don't go well with changing multipliers. Linking every down with an up, setting the multiplier at the end, or changing that input scheme entirely would be potential solutions

I see a few ways to fix this with various levels of ease of implementation, simplicity for future years, and flexibility for future scoring schemes.

My original idea behind the counter arrows as an input was an easily spammable button for things that higher-frequency than the other buttons, and easier to count back down if it gets over-clicked. 

I'm not sure if the counter arrows necessarily aligns with our goals for the UI of the judging or not, or ease for the judges. We could change them if we want.

**Quick rundown of current scoring:**
Every time a judge hits a button (or arrow) a new tuple gets added to the `Actions` table. It stores: timestamp, team, game #, scoring type, multiplier, up/down boolean, and deleted boolean

The scoring calculations count the # of entries for a scoring type at each multiplier level, subtracts `up - down` and then adds `count * multiplier` to the sum. 

This runs into problems if the judge:
counts up (or down) 
changes multiplier
counts down (or up)

This causes the down and up to not cancel each other out in the counts, and also be worth different amounts of points, so the points are incorrect. This should result in 0 points. Note that this year this can only happen to autonomous teams who switch to manual, but in future years we want to support fancy multipliers.

**Possible Solutions**
There are multiple different ways we could fix this with the counters as they are. I don't know what to do, since _I think a lot of the solutions I've come up with are too jank, and not easy for future years._ **TLDR:** 1 and 3 would be easy for this year. 4, 5, 6 maybe would be better for future years clarity

1. Every time the multiplier changes, we could have the server set the counters to 0

2. We could change the up/down format of the db tuples, and have the down arrows query for an up, and set it to deleted (this would lose our ability to go negative, but would make more sense to a judge (if the counter was counted up while on a multiplier, and then it was changed, going down to 0 would still properly remove the old db tuples)) (still not very visual for the judges)

3. We could have the multipliers for counters get updated when the game ends

4. We could change the way counters are counted in the db. Just have it update a single value, rather than record a new row for every button press

5. We don't use these counters, and we switch to type-in boxes and storing it as in 4

6. We don't use these counters, and we switch to buttons like the rest with delete buttons for every action.

**Comment by lpulley** [comment]
I think option 4 would be pretty good. If I'm correct, the only (but significant) downside would be that we wouldn't get an event-based list of actions, unless we had a separate table for "actions" and a "scores" table for the actual scores, or something like that. Options 5 and 6 also seem good. Which among 4, 5, and 6 would be easiest to do before competition this year?

**Comment by mgale456** [comment]
6 would be easiest, since it is already done, but we'll do 4/5, since I know Jacob has mentioned wanting to type stuff in, rather than use buttons. Shouldn't be too bad, I think.

**Comment by mgale456** [comment]
4/5 are done and work afaik. Only one database entry for them per game+team, and it stores the value. Other buttons are unaffected scoring-wise. They all get a 1 for the value, and it is summed now, rather than counted.
